### PR TITLE
better repalce_node fn

### PR DESCRIPTION
**Pull request by calad0i**
# Description

Make `ModelGraph.replace_node` method work with nodes work with replacing nodes, which itself/parents/children have multi-input/outputs.

#932 will be fixed with this PR.

This fix is rather long but keeps the previous behavior in most cases. A much shorter version will be simply doing

```python
        new_node.inputs = old_node.inputs
        new_node.outputs = old_node.outputs
```

but the final variable names will be different from the current version.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)

## Tests

Integrated with `test_repack_precision`, renamed to `test_repack_stream`

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: test/pytest/test_repack_stream.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from pathlib import Path
+
+import numpy as np
+import pytest
+from tensorflow import keras
+
+from hls4ml.converters import convert_from_keras_model
+
+# test_root_path = Path(__file__).parent
+test_root_path = Path('/tmp')
```
Comment (jmitrevs) [inline test/pytest/test_repack_stream.py:10]: This needs to be reverted.

**Comment by jmitrevs** [comment]
So the order of the nodes would stay the same, right? Is there a possibility to reorder the nodes? (I am not saying it is needed; just trying to understand.)

**Comment by jmitrevs** [comment]
I will tentatively put this as 0.8.1 milestone, but I would be fine deferring to 1.0.

**Comment by calad0i** [comment]
> So the order of the nodes would stay the same, right? Is there a possibility to reorder the nodes? (I am not saying it is needed; just trying to understand.)
> 

They should stay the same for the current implementation (copy in order w/ replacing a single k/v pair).

What do you mean by reorder in this case? Shuffling the nodes will break the graph in general I guess.

**Comment by jmitrevs** [comment]
I was thinking if there was a case where the new node expected the order of inputs to be different than the old node. It's more hypothetical, without a specific case in mind. ONNX nodes often have multiple inputs with specific roles assigned to specific inputs, but we usually transform the special inputs early.

**Comment by jmitrevs** [comment]
Do the tests ever produce a case when the length or the inputs or outputs is not one? I tried looking with a debugger and didn't see it in the tests, but I could have missed it.

File: test/pytest/test_repack_stream.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    out = keras.layers.Concatenate(name='concat')([r1, r2])
+    model = keras.Model([inp1, inp2], out)
+
+    hls_config = {'Model': {'Precision': 'ap_ufixed<8,8>', 'ReuseFactor': 1}, 'Strategy': strategy}
+    model_hls = convert_from_keras_model(
+        model,
+        io_type='io_stream',
+        backend='Vivado',
+        hls_config=hls_config,
+        output_dir=str(test_root_path / f'{backend}_{strategy}'),
```
Comment (jmitrevs) [inline test/pytest/test_repack_stream.py:52]: Generally the names have been of the style, hl4mlprj_repack_{backend}_{strategy}. The way it is currently, if you run pytest locally it's hard to refer back to the results.

File: test/pytest/test_repack_stream.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    layer_conf = {
+        'inp': {'Precision': 'fixed<20,10>'},
+        'reshape': {'Precision': 'fixed<20,10>'},
+        'conv': {'Precision': 'fixed<20,10>'},
+    }
+
+    hls_config = {'Model': {'Precision': 'fixed<2,1>', 'ReuseFactor': 1}, 'LayerName': layer_conf}
+
+    # Repack only happens in io_stream
+    model_hls = convert_from_keras_model(model, hls_config=hls_config, io_type='io_stream')
```
Comment (jmitrevs) [inline test/pytest/test_repack_stream.py:27]: Should name the output directory in the proper style.

**Comment by jmitrevs** [comment]
By the way, to answer my question about the replaced nodes having one input and output only in the test, I think that's true, but this fix properly handles the case when the node after the replaced node has multiple inputs, which the test demonstrates. So after properly setting the output directories, I think I can merge this PR and make 0.8.1.

**Review by jmitrevs — APPROVED** [review]
